src/main/java/apoc/util/Util.java

```diff
@@ -281,32 +282,68 @@ private static String handleRedirect(URLConnection con, String url) throws IOExc
 
     public static CountingInputStream openInputStream(String urlAddress, Map<String, Object> headers, String payload) throws IOException {
         StreamConnection sc;
+        InputStream stream;
+        if (urlAddress.contains(".zip") || urlAddress.contains(".tar") || urlAddress.contains(".tgz")) {
+            String[] tokens = urlAddress.split("!");
+            urlAddress = tokens[0];
+            String zipFileName = tokens[1];
+
+            sc = getStreamConnection(urlAddress,headers,payload);
+            stream = getFileStreamIntoCompressedFile(sc.getInputStream(), zipFileName);
+
+            return new CountingInputStream(stream, sc.getLength());
+        }
+
+        sc = getStreamConnection(urlAddress,headers,payload);
+        stream = getInputStream(sc,urlAddress);
+
+        return new CountingInputStream(stream, sc.getLength());
+    }
+
+    private static StreamConnection getStreamConnection(String urlAddress, Map<String, Object> headers, String payload) throws IOException {
         URL url = new URL(urlAddress);
         String protocol = url.getProtocol();
-        if(FileUtils.S3_PROTOCOL.equalsIgnoreCase(protocol)) {
-            sc = FileUtils.openS3InputStream(url);
-        } else if(FileUtils.HDFS_PROTOCOL.equalsIgnoreCase(protocol)) {
-            sc = FileUtils.openHdfsInputStream(url);
+        if (FileUtils.S3_PROTOCOL.equalsIgnoreCase(protocol)) {
+            return FileUtils.openS3InputStream(url);
+        } else if (FileUtils.HDFS_PROTOCOL.equalsIgnoreCase(protocol)) {
+            return FileUtils.openHdfsInputStream(url);
         } else {
-            sc = readHttpInputStream(urlAddress, headers, payload);
+            return readHttpInputStream(urlAddress, headers, payload);
         }
+    }
 
-        String encoding = sc.getEncoding();
+    private static InputStream getInputStream(StreamConnection sc, String urlAddress) throws IOException {
         InputStream stream = sc.getInputStream();
+        String encoding = sc.getEncoding();
+
         if ("gzip".equals(encoding) || urlAddress.endsWith(".gz")) {
-            stream = new GZIPInputStream(stream);
+             return new GZIPInputStream(stream);
         }
         if ("deflate".equals(encoding)) {
-            stream = new DeflaterInputStream(stream);
+            return new DeflaterInputStream(stream);
         }
-        return new CountingInputStream(stream, sc.getLength());
```

Comment:
as you remove CountingInputStream here, did you fix that in all places that call: `getInputStream` ?

Reply:
this is old, I done a small refactoring, i put I comment where is the new return countingInputStream, is inside the openInputStream method as before